File: Two_Pointers/C++/167_two_sum_2.cpp
/*
* RUNTIME: 9 ms - Beats 65.33% of submissions
* MEMORY - 17.96 MB - Beats 5.28% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
using namespace std;


/* Thoughtprocess:
* 1. Initiate two pointers, one at the left and one on the right
* 2. While left < right, check if the sum of the two pointers is equal to the target
* 3. If the sum is less than the target, move the left pointer to the right
* 4. If the sum is greater than the target, move the right pointer to the left
* 
*/


class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int left = 0, right = numbers.size()-1;
        vector<int> result;
        while (left < right)
        {
            if (numbers[left] + numbers[right] == target)
            {
                result.push_back(left+1);
                result.push_back(right+1);
                return result;
            }
            else if (numbers[left] + numbers[right] < target)
                left++;
            else
                right--;
        }
        return result;
    }
};

File: Two_Pointers/Python/167_two_sum_2.py
# RUNTIME: 99 ms, faster than 95.50% of python3 Submissions. 
# MEMORY USAGE: 17.38 MB, less than 65.45% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Set left and right borders
# 2. if sum of left and right is equal we have the suolution
# 3. if sum of left and right is less than target, we need to increase left
# 4. if sum of left and right is greater than target, we need to decrease right
# 5. continue until solution is found or left is greater than right
# 6. Because there is a solution, we don´t need the last return, but it is good to have it




class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left, right = 0, len(numbers) - 1
        
        while left < right:
            if numbers[left] + numbers[right] == target:
                return [left + 1, right + 1]
            elif numbers[left] + numbers[right] < target:
                left += 1
            else:
                right -= 1
        return [-1, -1]